File: proj2md-js/lib/discover.js
'use strict';
const fs = require('fs');
const path = require('path');
const D = require('./defaults.js');
const { fnmatch } = require('./util.js');
function _matchAny(relPosix, nameLower, patterns) {
  for (const pat of patterns) {
    const patL = String(pat).toLowerCase();
    if (fnmatch(nameLower, patL) || fnmatch(relPosix, patL)) return true;
  }
  return false;
}
/** 判断某目录是否可能被 --include-pattern 覆盖（让强制包含穿透「隐藏目录忽略」）。 */
function _dirMayBeIncluded(dirRel, patterns) {
  for (const pat of patterns) {
    const prefix = String(pat).toLowerCase().split('*')[0];
    if (!prefix) return true;
    if (dirRel.toLowerCase().startsWith(prefix)) return true;
    if (prefix.startsWith(dirRel.toLowerCase() + '/')) return true;
  }
  return false;
}
function _realish(p) { try { return fs.realpathSync(p); } catch (e) { return p; } }
/**
 * 遍历项目收集候选文件，返回 { found: [{p, rel}], prunedHidden: [relDir] }。
 * 目录剪枝顺序：include-pattern 覆盖 > 隐藏目录忽略 > 目录黑名单。
 * rel 统一为 posix 风格（"/"分隔），与输出格式一致。
 */
function discover(cfg) {
  const root = cfg.root;
  const outAbs = _realish(path.resolve(cfg.output));
  const cfgAbs = cfg.configPath ? _realish(path.resolve(cfg.configPath)) : null;
  // 与 Python 版对齐：discover / cli / 运行入口三者都排除，
  // 否则把包目录当项目跑时会把自己的源码拼进合集。
  const selfAbs = new Set([
    _realish(__filename),
    _realish(path.join(__dirname, 'cli.js')),
  ]);
  if (process.argv[1]) selfAbs.add(_realish(path.resolve(process.argv[1])));
  // 目录黑名单统一小写后拆分：精确匹配走 Set（O(1)），通配模式（*.egg-info）单独列表。
  const excludeDirs = new Set([...cfg.excludeDirs].map((p) => String(p).toLowerCase()));
  const excludeGlobs = [...excludeDirs].filter((p) => p.includes('*'));
  const excludeExact = new Set([...excludeDirs].filter((p) => !p.includes('*')));
  const found = [];
  const prunedHidden = [];
  const walk = (dir, relBase) => {
    let entries;
    try { entries = fs.readdirSync(dir, { withFileTypes: true }); }
    catch (e) { return; }
    const dirs = [], files = [];
    for (const e of entries) {
      if (e.isSymbolicLink()) {
        // 链接一律不递归（Python 侧 os.walk 不跟随符号链接，并在 discover 里
        // 显式跳过 junction——跟随会让子树被收录两遍，成环时还会无限递归）。
        // 文件链接照常收录，与 Python 一致；断链/无权限时由读取阶段报「无法读取」。
        let st = null;
        try { st = fs.statSync(path.join(dir, e.name)); } catch (err) { st = null; }
        if (st && st.isDirectory()) continue;
        files.push(e.name);
        continue;
      }
      if (e.isDirectory()) dirs.push(e.name);
      else if (e.isFile()) files.push(e.name);
    }
    dirs.sort(); files.sort();
    // ── 目录剪枝 ──
    const kept = [];
    for (const d of dirs) {
      const relDir = relBase ? relBase + '/' + d : d;
      if (_dirMayBeIncluded(relDir, cfg.includePatterns)) { kept.push(d); continue; }
      if (cfg.excludeHidden && d.startsWith('.')) { prunedHidden.push(relDir); continue; }
      const dLower = d.toLowerCase();
      // 支持通配目录名（如 "*.egg-info"）
      if (excludeExact.has(dLower)) continue;
      if (excludeGlobs.some((pat) => fnmatch(dLower, pat))) continue;
      kept.push(d);
    }
    // ── 当前目录文件（与 os.walk 顺序一致：先本层文件，再递归子目录）──
    for (const fn of files) {
      const p = path.join(dir, fn);
      const rel = relBase ? relBase + '/' + fn : fn;
      const nameL = fn.toLowerCase();
      const pa = _realish(p);
      if (pa === outAbs || (cfgAbs && pa === cfgAbs) || selfAbs.has(pa)) continue;
      const includedOverride = _matchAny(rel, nameL, cfg.includePatterns);
      if (cfg.excludeFiles.has(nameL) && !includedOverride) continue;
      if (_matchAny(rel, nameL, cfg.excludePatterns) && !includedOverride) continue;
      const dot = fn.lastIndexOf('.');
      const ext = dot > 0 ? fn.slice(dot + 1).toLowerCase() : '';
      if (!(cfg.anyText || cfg.exts.has(ext) || D.DEFAULT_FILENAMES.has(nameL))) {
        if (!includedOverride) continue;
      }
      found.push({ p, rel });
    }
    for (const d of kept) walk(path.join(dir, d), relBase ? relBase + '/' + d : d);
  };
  walk(root, '');
  return { found, prunedHidden };
}
module.exports = { discover };

File: proj2md-py/proj2md/discover.py
"""proj2md 文件发现、编码识别读取与记录构建。"""
from __future__ import annotations
import fnmatch
import os
from pathlib import Path

from .config import Config, FileRec
from .defaults import CONFIG_MANIFESTS, DEFAULT_FILENAMES, ENTRY_EXTS, ENTRY_STEMS
from .i18n import t
from .util import fmt_size, lang_of

# ─────────────────────────── 文件发现与读取 ───────────────────────────
def _match_any(rel_posix: str, name_l: str, patterns) -> bool:
    for pat in patterns:
        pat_l = str(pat).lower()
        if fnmatch.fnmatch(name_l, pat_l) or fnmatch.fnmatch(rel_posix, pat_l):
            return True
    return False
def _dir_may_be_included(dir_rel: str, patterns) -> bool:
    """判断某个目录是否可能被 --include-pattern 覆盖（用于让强制包含
    穿透「隐藏目录忽略」）。取每个 pattern 第一个 * 之前的字面前缀：
      - 前缀为空（如 "*"、"*.md"）→ 可能覆盖一切目录 → True
      - 目录路径以前缀开头（如 ".github" vs ".github/*"）→ True
      - 前缀以「目录/」开头（目录是 pattern 覆盖范围的祖先）→ True
    宁可放宽（多遍历再逐文件判断），不可误剪。"""
    for pat in patterns:
        prefix = str(pat).lower().split("*")[0]
        if not prefix:
            return True
        if dir_rel.lower().startswith(prefix):
            return True
        if prefix.startswith(dir_rel.lower() + "/"):
            return True
    return False
_JUNCTION_TAG = 0xA0000003   # IO_REPARSE_TAG_MOUNT_POINT（Windows 目录 junction）
def _is_dir_link(path: Path) -> bool:
    """目录形式的链接：Unix/Windows 符号链接，以及 Windows junction。

    Node 侧 fs.Dirent.isSymbolicLink() 对 junction 同样返回 true，这里保持一致。
    跟随这类目录会让同一棵子树被收录两遍（junction 还能指回祖先目录，
    使遍历无限递归），所以两端都直接跳过、不递归。"""
    try:
        if os.path.islink(path):
            return True
        return getattr(os.lstat(path), "st_reparse_tag", 0) == _JUNCTION_TAG
    except OSError:
        return False
def _dir_excluded(name_l: str, patterns) -> bool:
    """目录黑名单匹配：大小写不敏感，支持 * / ? 通配（如 *.egg-info）。"""
    for pat in patterns:
        pat_l = str(pat).lower()
        if name_l == pat_l or fnmatch.fnmatchcase(name_l, pat_l):
            return True
    return False
def discover(cfg: Config):
    """遍历项目收集候选文件。返回。
    目录剪枝顺序：include-pattern 覆盖 > 隐藏目录忽略 > 目录黑名单。"""
    root = cfg.root
    out_abs = cfg.output.expanduser().resolve()
    cfg_abs = cfg.config_path.resolve() if cfg.config_path else None
    # 与 Node 版对齐：排除 discover / cli / __main__ 三份自身驱动源文件，
    # 否则把包目录当项目跑（回归验证的常见做法）时会把自己的源码拼进合集。
    # 用包目录下的固定文件名而非 sys.modules：结果确定，且绝不会误伤用户自己的模块。
    pkg_dir = Path(__file__).resolve().parent
    self_abs = {
        Path(__file__).resolve(),
        pkg_dir / "cli.py",
        pkg_dir / "__main__.py",
    }
    found, pruned_hidden = [], []
    for dirpath, dirnames, filenames in os.walk(root):
        kept = []
        for d in sorted(dirnames):
            rel_dir = (Path(dirpath) / d).relative_to(root).as_posix()
            # 目录链接（符号链接 / junction）一律不递归，且不受 include-pattern 影响：
            # 对应 Node 侧在 readdir 阶段就把它排除出 dirs，两端行为需一致。
            if _is_dir_link(Path(dirpath) / d):
                continue
            # 强制包含规则优先：可能被 include-pattern 覆盖的目录一律不剪
            if _dir_may_be_included(rel_dir, cfg.include_patterns):
                kept.append(d)
                continue
            # 隐藏目录：以 . 开头且开启忽略 → 整目录剪掉
            if cfg.exclude_hidden and d.startswith("."):
                pruned_hidden.append(rel_dir)
                continue
            # 目录黑名单（支持 *.egg-info 这类通配目录名）
            if _dir_excluded(d.lower(), cfg.exclude_dirs):
                continue
            kept.append(d)
        dirnames[:] = kept
        for fn in sorted(filenames):
            p = Path(dirpath) / fn
            rel = p.relative_to(root)
            rel_posix = rel.as_posix()
            name_l = fn.lower()
            try:
                pa = p.resolve()
            except OSError:
                pa = p
            if pa in self_abs or pa == out_abs or (cfg_abs and pa == cfg_abs):
                continue
            included_override = _match_any(rel_posix, name_l, cfg.include_patterns)
            if name_l in cfg.exclude_files and not included_override:
                continue
            if _match_any(rel_posix, name_l, cfg.exclude_patterns) and not included_override:
                continue
            ext = p.suffix.lower().lstrip(".")
            if not (cfg.any_text or ext in cfg.exts or name_l in DEFAULT_FILENAMES):
                if not included_override:
                    continue
            found.append((p, rel))
    return found, pruned_hidden
def read_text(p: Path):
    """自动识别编码读取文本；返回。二进制返回。"""
    try:
        raw = p.read_bytes()
    except OSError as e:
        return None, None, t("read_fail", cls=e.__class__.__name__)
    if b"\x00" in raw:
        return None, None, t("looks_binary")
    for enc in ("utf-8-sig", "utf-8", "gbk", "big5", "latin-1"):
        try:
            return raw.decode(enc), enc, None
        except (UnicodeDecodeError, LookupError):
            continue
    return None, None, t("undecodable")
def file_priority(p: Path) -> int:
    name_l = p.name.lower()
    if name_l.startswith("readme"):
        return 0
    if name_l in CONFIG_MANIFESTS or name_l in (".gitignore", ".dockerignore", ".editorconfig"):
        return 1
    if p.stem.lower() in ENTRY_STEMS and p.suffix.lower() in ENTRY_EXTS:
        return 2
    if p.stem.lower() in ("config", "settings"):
        return 2
    return 3
def order_key(item):
    p, rel = item
    return (file_priority(p), rel.as_posix().lower())
def build_records(cfg: Config, candidates):
    records, skipped = [], []
    total = 0  # ★ 修复（P0-1）：预算按 UTF-8 字节计，与 --max-file-kb 的字节口径一致
               #   （原按字符数计，中文项目实际产物体积可达预算的 ~3 倍）
    budget = int(cfg.max_total_kb * 1024) if cfg.max_total_kb else 0
    for p, rel in candidates:
        try:
            size = p.stat().st_size
        except OSError as e:
            skipped.append((rel.as_posix(), t("skip_unreadable", cls=e.__class__.__name__)))
            continue
        if cfg.max_file_kb and size > cfg.max_file_kb * 1024:
            skipped.append((rel.as_posix(),
                            t("skip_too_large", kb=cfg.max_file_kb, size=fmt_size(size))))
            continue
        text, enc, err = read_text(p)
        if text is None:
            skipped.append((rel.as_posix(), t("skip_read_err", err=err)))
            continue
        text = text.replace("\r\n", "\n").replace("\r", "\n")
        if not text.endswith("\n"):
            text += "\n"
        orig_lines = len(text.splitlines())
        truncated = False
        if cfg.max_file_lines and orig_lines > cfg.max_file_lines:
            keep = cfg.max_file_lines
            text = "\n".join(text.split("\n")[:keep])
            if not text.endswith("\n"):
                text += "\n"
            text += t("truncated_note", orig=orig_lines, keep=keep)
            truncated = True
        if not text.strip():
            text = t("empty_file")
        nb = len(text.encode("utf-8"))  # ★ 按最终写入正文的字节数计（截断/换行归一之后）
        if budget and records and total + nb > budget:
            skipped.append((rel.as_posix(), t("skip_over_budget")))
            continue
        records.append(FileRec(
            rel=rel, abspath=p, language=lang_of(p), encoding=enc,
            content=text, lines=len(text.splitlines()), chars=len(text),
            nbytes=size, truncated=truncated, orig_lines=orig_lines,
        ))
        total += nb
    return records, skipped
